flybywiresim/a32nx
# feat(bleed): replace default bleed system

## Summary of Changes

Implementation of the Bleed system written in Rust.

As a complete compressible fluid flow simulation is not feasibly implementable, we settle for a basic simulation assuming isentropic flow.

In code, this is implemented using basic components: containers and valves. A container contains some amount of fluid (in this case air). The container is responsible for storing thermodynamic quantities of the fluid stored in it, i.e pressure and temperature. 
A valve on the other hand is responsible for transferring fluids between containers. The valve removes some amount of air from one container and adds it to the other container. Consequently, both containers update their macroscopic quantities using equations for an isentropic process. There are more smaller components to support the rest of the system.

The implementation of the system can be roughly divided into the subtasks listed below. Note that tasks marked as complete may still require tuning.
- [x] Basic fluid simulation elements (valves, containers, etc.)
- [x] Engine related systems
  - [x] Basic engine model (i.e pressure in the compression chambers at different altitudes/speeds)
  - [x] HP / IP valve switching logic
  - [x] PRV regulation logic
  - [x] Engine starter valve
  - [x] Fan air valve
  - [x] Precooler
- [x] APU
  - [x] APU bleed valve
- [x] Overhead panel integration
- [x] Cross bleed valve
- [x] Other bleed air consumers
  - [x] A very basic implementation of the packs.
  - [x] Hydraulic reservoirs
- [x] Connect everything to proper electrical busses

Not addressed in this PR:
- Cabin pressurization
- Packs

Known issues & limitations:
- Nothing of the new bleed system is shown on the ECAM yet. All the variables are visible in the ModelBehaviours screen. I will most likely do this in a separate PR.
- There are a lot of parameters that still need to be tuned to make the pressure and temperature values shown more accurate. This is quite hard to do due to lack of references.
- While the bleed system has the logic for engine starting in place, this is not actually connected to the engines yet. This also means that crossbleed engine starts are not yet possible
- All pressure containers are initialized at 1 standard atmosphere (14.7 psi). This is ok for now since the system reaches the proper steady-state pretty quickly.
- Heat convection is not yet modelled, only heat conduction.

## Screenshots (if necessary)



## References

Tons of them




## Additional context

A slightly outdated overview of the bleed system in the A320.
![image](https://user-images.githubusercontent.com/22713769/145071448-c1de3b60-d30a-408a-9caf-8e0a7450be18.png)



Discord username (if different from GitHub): BlueberryKing#6641

## Testing instructions

To test this PR, you need to have the ModelBehaviours window open in MSFS. That's how you can check the values of the pneumatic SimVars.

- Start cold & dark, do not turn anything on for now.
- Open the debugger, type in "PNEU" to see all the bleed system-related variables.
- Make sure all pressure values are around atmospheric pressure (14.7 psi). These do not change with the ambient pressure for now.
- Make sure all of the temperatures are at around 15 °C.
- Since the aircraft does not have any power, doing the following should not have any effect on valves.
  - Try clicking the PACK 1 and PACK 2 pushbuttons.
  - Try clicking the ENG 1 BLEED and ENG 2 BLEED pushbuttons
  - Turn the crossbleed valve control knob
- Start up APU
- Make sure APU bleed valve is closed, turn on the APU bleed. Check:
  - The APU bleed valve should now be open. (Look for 'apu' where you currently have 'pneu' in the variables window and find `A32NX_APU_BLEED_AIR_VALVE_OPEN`)
  - The Crossbleed valve should also be open
  - Check that the pressure regulating valves are closed.
  - Observe both `PRECOOLER_OUTLET_TEMPERATURE` and `PRECOOLER_INLET_PRESSURE` rising. It is expected that pressures also rise in other sections of the system. Make sure that there is no big increase in the `HP_PRESSURE` or `IP_PRESSURE`.
- Turn the crossbleed valve control knob to the SHUT position. Make sure that pressure on the number 2 system drops by a bit.
- Turn the knob back to AUTO and see how the pressure comes back.
- For both sides, turn off the PACK pushbutton. Observe:
  - The pack flow dropping to zero on that side.
- Start the engines.
  - You should see HP pressure and temperature increase with N2. 
  - IP pressure and temperature should also increase once N1 is spinning.
- Turn off the APU bleed. Observe:
  - The Pressure regulating valve opening (`A32NX_PNEU_ENG_1_PR_VALVE_OPEN` and `A32NX_PNEU_ENG_2_PR_VALVE_OPEN`)
  - Precooler outlet pressures should settle at around 46 psi.
- Do a runup. On one or both engines
  - Slowly increase thrust and observe pressures and temperatures increasing in HP and IP.
  - If HP valve was open previously, it should close at some point during the runup and the IP valve should open instead.
  - The precooler inlet and outlet pressure and temperatures should only increase slightly.
- Taxi and takeoff
  - Observe how the precooler outlet pressures stay around 46 psi.
  - Play around with the system. Make sure no pressure or temperature ever has a value of `nan`, `inf` or `-inf`.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
